docs/troubleshooting.rst

```diff
@@ -277,6 +277,127 @@ the log level of the command to the DEBUG level and the ``-X`` option shows stac
 
    $ inmanta -vvv export -X
 
+
+Compilation fails
+=================
+
+In rare situations, the compiler might fail with a ``List modified after freeze`` or an
+``Optional variable accessed that has no value`` error, even thought the model is syntactically correct. The following
+sections describe why this error occurs and what can be done to make the compilation succeed.
+
+Reason for compilation failure
+------------------------------
+
+When the compiler runs, it cannot know upfront how many elements will be added to a relationship. At some stages of the
+compilation process the compiler has to guess which relations are completely populated in order to be able to continue
+the compilation process. Heuristics are being used to determine the correct order in which relationships can be
+considered completely populated. In most situation these heuristics work well, but in rare situations the compiler
+makes an incorrect decision and considers a relationship to be complete while it isn't. In those situation the compiler
+crashes with one of the following exception:
+
+* ``List modified after freeze``: This error occurs when a relationship with an upper arity larger than one was
+  considered complete too soon.
+* ``Optional variable accessed that has no value``: This error occurs when a ``[0:1]`` relationship was considered
+  complete too soon.
+
+The following sections provide information on how this issue can be resolved.
+
+Relationship precedence policy
+------------------------------
+
+.. warning::
+
+    The relationship precedence policy feature is currently experimental
```

Comment:
@wouterdb do we still consider it bad practice to rely on this feature as anything other than a temporary workaround? Is it always better and feasible to solve the issue by removing ambiguity in the model?

Reply:
I consider it a last resort for now.

Reply:
I would indeed give a warning that this can easily make things worse. Something like:

```
The inmanta compiler is very good at determining in which order it should evaluate the orchestration model. Unfortunately in very complex models it might not be able to do this. In that case you can give the compiler some instruction by providing it with relationship precedence rules.

This is a very powerful tool because you can override all the intelligence in the compiler. This means that if you provide the correct rule it will fix the compilation. If you provide a wrong rule it can make this even worse. However, it can never make the orchestrator compile incorrect results.
```